File: BackEnd/src/data/IMappers/IPlantSpeciesMapper.ts
/**
 * @file IPlantSpeciesMapper.ts
 * @description Interfaz del mapper de especies de plantas. Contrato de conversión PlantSpeciesDocument ↔ PlantSpecies ↔ PlantSpeciesResponseDTO.
 * @module Plants
 * @layer Data
 */

import type { PlantSpecies } from '../../domain/entities/PlantSpecies.js';
import type { PlantSpeciesDocument } from '../datasources/mongodb/models/PlantSpeciesModel.js';
import type { PlantSpeciesResponseDTO } from '../../domain/dtos/plants/plant-species-response.dto.js';

/**
 * Contrato del mapper de especies de plantas.
 */
export interface IPlantSpeciesMapper {
  /**
   * Convierte un documento MongoDB a entidad de dominio.
   *
   * @param doc — Documento de la colección 'plant_species'.
   * @returns Entidad PlantSpecies.
   */
  toEntity(doc: PlantSpeciesDocument): PlantSpecies;

  /**
   * Convierte una entidad de dominio a documento MongoDB (sin _id).
   *
   * @param entity — Entidad PlantSpecies.
   * @returns Documento listo para insertar/actualizar.
   */
  toDocument(entity: PlantSpecies): Omit<PlantSpeciesDocument, '_id'>;

  /**
   * Convierte una entidad de dominio al DTO de respuesta para el cliente.
   *
   * @param entity — Entidad PlantSpecies.
   * @returns DTO serializable para la respuesta HTTP.
   */
  toResponseDTO(entity: PlantSpecies): PlantSpeciesResponseDTO;
}

File: FrontEnd/lib/data/i_mappers/i_plant_species_mapper.dart
/// @file i_plant_species_mapper.dart
/// @description Interfaz del mapper de especies. Contrato PlantSpeciesModel ↔ PlantSpecies.
/// @module Plants
/// @layer Data
library;

import '../../domain/entities/plant_species.dart';
import '../models/plant_species_model.dart';

/// Contrato de conversión entre el modelo de serialización y la entidad de dominio.
///
/// [injectable] registrar en container.dart como singleton.
abstract interface class IPlantSpeciesMapper {
  /// Convierte un modelo de serialización a entidad de dominio.
  PlantSpecies toEntity(PlantSpeciesModel model);

  /// Convierte una entidad de dominio a modelo de serialización.
  PlantSpeciesModel toModel(PlantSpecies entity);
}
